### DISCUSSION TITLE
Empty Snapshot when implementing own Serializer

### DISCUSSION THREAD

**Discussion by ClaudioZesiger** [origin-energy/java-snapshot-testing#164] [Q&A] [answered]
Hello,
I implemented my own Serializer with Jsonb. The snapshot is written well into the snap-file. But when comparing the snapshot to another snapshot, the body of the existing snapshot from the file is always an empty string.

This is my apply method
`
@Override
public Snapshot apply(Object o, SnapshotSerializerContext context) {
  String json = this.jsonb.toJson(o);
  return context.toSnapshot(json);
}
` 
The jsonb object is initialized.

I saw that the regex in the implementation does not support line breaks. File Snapshot.java.
`
String regex = "^(?<name>.*?)(\\[(?<scenario>.*)\\])?=(?<header>\\{.*\\})?(?<snapshot>(.*)$)";
`
But normally any JSON should be formatted for readability.

Any ideas about this?

Regards
C

**Comment by jackmatt2** [comment]
The regex is using [DOTALL](https://github.com/origin-energy/java-snapshot-testing/blob/master/java-snapshot-testing-core/src/main/java/au/com/origin/snapshots/Snapshot.java#L33) which means it will also match line breaks.

I'd but a debug point into the deserialization and check why it's returning an empty string.

**Reply by ClaudioZesiger** [reply]
Hey,
I already did this. I debugged the whole deserialization. The problem is, that the body is not recognized by the regex.
When i copy the JSON in the snapshot file and the regex to a regex webpage online like regex101.com, I can see that the body is recognized as part of the header.

Here is an example with a simple TestObject which has an integer i and a string s inside.

Content of the snapshot
any.package.SnapshotTest.testSnapshotCreation={
  "typeName": "any.package.SnapshotTest$TestObject"
}{
    "i": 100,
    "s": "Snapshot Test"
}

Everything after the equal sign is taken as the header.
This is why it returns an empty string in the body.

Any ideas?

Regards
C

**Reply by jackmatt2** [reply]
The snapshot needs Square brackets around it. See this in the Jackson implementation 

https://github.com/origin-energy/java-snapshot-testing/blob/master/java-snapshot-testing-plugin-jackson/src/main/java/au/com/origin/snapshots/jackson/serializers/DeterministicCollectionModule.java#L48

**Reply by jackmatt2** [reply]
Also hard coded in string serializer

https://github.com/origin-energy/java-snapshot-testing/blob/master/java-snapshot-testing-core/src/main/java/au/com/origin/snapshots/serializers/v1/ToStringSnapshotSerializer.java#L20

**Reply by ClaudioZesiger** [reply]
I cannot understand why I need square brackets around the body. They are not part of the regex, except for the scenario.

Would it be enough to change this?
`
@Override
public Snapshot apply(Object o, SnapshotSerializerContext context) {
    String json = "[\n" + this.jsonb.toJson(o) + "\n]";
    return context.toSnapshot(json);
}
`

**Reply by jackmatt2** [reply]
Yeah, actually it could be anything I think, just not a closing curly brace. I'd stick with square brackets for now though.

It's a legacy artifact from the first iterations whereby one could pass a varargs and it was rendered as an array. I needed to maintain backward compatibility so was forced to maintain the Square brackets in the serializer. 

What you have above is fine

**Reply by ClaudioZesiger** [reply]
I'll try it and give you feedback next week.

**Reply by ClaudioZesiger** [reply]
Good morning,
I have tried the solution discussed above. It was not successful.
The square brackets didn't change anything.

Just to be concrete, this is the content of my snapshot file when generating it with Jsonb Serializer.

tech.bison.modultests.util.snapshot.SnapshotTest.testSnapshotCreation={
  "typeName": "tech.bison.modultests.util.snapshot.SnapshotTest$TestObject"
}[
{
    "i": 100,
    "s": "Snapshot Test"
}
]

Name is read correct by the regex when debugging.
The part from the first curly bracket (after the equal sign) to the last curly bracket is read as the header, which is wrong in my opinion.

Some more ideas how to serialize and extend the java-snapshot-testing with Jsonb?

Regards
C

**Comment by jackmatt2** [comment]
@ClaudioZesiger There indeed appears to be an issue with headers + JSON body when parsed using the REGEX.  Should be fixed in the upcoming 4.0.8 release

The Maven release is being flakey today so I will try it again tomorrow.

**Comment by jackmatt2** [comment, marked as answer]
Can you update to 4.0.8 - this should fix the issue you are having.  I don't think the square brackets are required either.
